### DISCUSSION TITLE
LSP

### DISCUSSION THREAD

**Discussion by Peyton-Spencer** [gpuweb/gpuweb#4315] [Ideas/Feedback]
WGSL has limited LSP options; it would be great to have an LSP of the same quality as rust-analyzer and gopls. It will 100x the WGSL dev experience.

**Comment by greggman** [comment]
I assume you mean: https://microsoft.github.io/language-server-protocol/

not saying someone shouldn't do it, maybe you?, but think it's harder than it looks. I expect lots of devs do various string manipulation to assemble WGSL. If you have your own preprocessor then you'd need an lsp that understands it for completion to work. rust and go have rules for how code is assembled, WGSL does not. It's left to the developer to do whatever they want via string assembly.

**Comment by dneto0** [comment]
The grammar for WGSL is machine-readable.  See https://github.com/gpuweb/gpuweb/blob/main/wgsl/syntax.bnf 
There are subtleties including:
- block comments are nestable; so you need a customized lexer to do that perfectly
- there's a prepass to disambiguate template  list start and end
- parsing and tokenization are interleaved (at least canonically; you don't have to implement it that way)

And then there's syntax checking and overload resolution;  that's fairly standard, with some parameterizations.

It's all doable. But like @greggman suggests, it's not the responsibility of W3C or its members to make that tooling.  It seems like a ripe opportunity for the wider community to produce this.

**Comment by Peyton-Spencer** [comment]
I appreciate your response, thank you for clarifying. Bevy has its own preprocessor for example. I have a better understanding of the complexity now.

As I don't have experience with LSP, my best bet is to contribute to [WGSL-analyzer](https://github.com/wgsl-analyzer/wgsl-analyzer) and attempt to solve the preprocessor issues I encounter.
